util/applications/src/iScience/generate_reads_refseq.cpp

```diff
@@ -162,78 +139,151 @@ void initialise_argument_parser(sharg::parser & parser, cmd_arguments & argument
-    }
-
     std::ifstream istrm{arguments.bin_file_path};
     std::string line;
     sharg::input_file_validator validator{};
+    std::filesystem::path bin_path{};
 
-    size_t sum_of_weights{};
     while (std::getline(istrm, line))
     {
-        if (!line.empty())
+        if (line.empty())
+            continue;
+
+        auto const & [path_span, weight_span] = [&line]()
         {
-            auto tab = std::find(line.begin(), line.end(), '\t');
+            auto const tab = std::ranges::find(line, '\t');
+            auto const end = line.end();
 
-            // parse file path
-            std::filesystem::path bin_path{line.begin(), tab};
-            validator(bin_path);
-            arguments.bin_path.push_back(std::move(bin_path));
+            std::span path_span{line.begin(), tab};
+            std::span weight_span{std::ranges::next(tab, 1, end), end};
 
-            // parse weight if given
-            if (tab != line.end())
-            {
-                ++tab;
-                size_t tmp{};
-                std::from_chars(&(*tab), &line[line.size() - 1], tmp);
-                sum_of_weights += tmp;
-                arguments.number_of_reads_per_bin.push_back(tmp); // initialise with weight
-            }
+            return std::make_tuple(path_span, weight_span);
+        }();
+
+        bin_path.assign(path_span.begin(), path_span.end());
+        validator(bin_path);
+        arguments.bin_path.push_back(bin_path);
+
+        if (arguments.weight_mode != weight::from_weight_column)
+            continue;
+
+        if (!weight_span.empty())
+        {
+            arguments.number_of_reads_per_bin.emplace_back();
+            std::from_chars(weight_span.data(),
+                            weight_span.data() + weight_span.size(),
+                            arguments.number_of_reads_per_bin.back());
+        }
+        else
+        {
+            throw sharg::parser_error{"No weight given for bin: " + bin_path.string()};
         }
     }
+}
+
+void infer_weights_from_kmer_counts(cmd_arguments & arguments)
+{
+    size_t const number_of_bins = arguments.bin_path.size();
+    arguments.number_of_reads_per_bin.resize(number_of_bins);
+    raptor::file_reader<raptor::file_types::sequence> reader{seqan3::ungapped{32u}, 32u};
+
+#pragma omp parallel for schedule(dynamic) num_threads(arguments.threads)
+    for (size_t i = 0; i < number_of_bins; ++i)
+    {
+        // could also be defined outside the loop and then be used with the pragma directtive `private(kmer_set)`
+        // and then be cleared at the end of the loop
+        // However, this would increase the maximum memory usage, as the set would stay at the maximum size of all
+        // bins that are processed by this thread.
+        robin_hood::unordered_flat_set<uint64_t> kmer_set{};
+        reader.hash_into(arguments.bin_path[i], std::inserter(kmer_set, kmer_set.begin()));
+        arguments.number_of_reads_per_bin[i] = kmer_set.size();
+    }
+}
```

Comment:
This is very tricky for large data, as using many threads will most likely use too much RAM. What about using hll sketches? `hibf::sketch::compute_sketches` the function is called I think. The estimate is enough to generate reads. If someone really wants kmer counts, they can use the column option.